File: plugins/openclaw/tools/graph_traverse.ts
/**
 * Nexus Memory — Knowledge Graph Tools for OpenClaw
 *
 * Provides graph traversal, entity search, subgraph, and related-facts queries.
 * Uses the shared Qdrant collection via REST API (same as other OpenClaw tools).
 */

import { Type } from "@sinclair/typebox"
import type { OpenClawPluginApi } from "openclaw/plugin-sdk"
import type { QdrantClient } from "../lib/qdrant-client.ts"
import type { NexusConfig } from "../lib/config.ts"
import { clampInt } from "../lib/num.ts"
import { isActiveEdge, normalizeEdge } from "../lib/edge.ts"
import { bfsEdges } from "../lib/bfs.ts"
import { log } from "../logger.ts"

// H123: hard bounds for the graph tools, mirrored by their Typebox schemas.
const TRAVERSE_DEPTH_DEFAULT = 3
const TRAVERSE_DEPTH_MAX = 5
const ENTITIES_LIMIT_DEFAULT = 50
const ENTITIES_LIMIT_MAX = 500
const SUBGRAPH_DEPTH_DEFAULT = 2
const SUBGRAPH_DEPTH_MAX = 5

// W31-18: relations that may enter the public result contract. Mirrors the
// canonical `nexus.graph.schema.EdgeRelation` values — the whitelist the
// outgoing (normalizeEdge) and incoming validation gates share.
const KG_RELATIONS: readonly string[] = [
  "supersedes",
  "contradicts",
  "supports",
  "alternative_to",
  "depends_on",
  "references",
  "installed_at",
  "connected_to",
  "manages",
  "runs_on",
  "part_of",
  "owns",
  "located_at",
  "depends_on_service",
  "uses",
  "provides",
  "controls",
]

/**
 * Validate one INCOMING edge (a `findIncomingEdges` result) before it enters
 * the public `{ fact_id, relation, edge_id }` contract.
 *
 * W31-18(a): the incoming branch used to push raw payload fields as casts —
 * mirroring the H125 gate the outgoing branch already runs through
 * (`normalizeEdge`): fact_id must be a non-empty string, relation must be a
 * non-empty whitelisted string, edge_id degrades to "" (never `undefined`).
 */
function validateIncomingEdge(
  edge: { source_id?: unknown; relation?: unknown; edge_id?: unknown } | null | undefined,
): { fact_id: string; relation: string; edge_id: string } | null {
  if (!edge || typeof edge !== "object") return null
  const factId = edge.source_id
  if (typeof factId !== "string" || factId === "") return null
  const relation = edge.relation
  if (typeof relation !== "string" || relation === "") return null
  if (!KG_RELATIONS.includes(relation)) return null
  const edgeId =
    typeof edge.edge_id === "string" && edge.edge_id !== "" ? edge.edge_id : ""
  return { fact_id: factId, relation, edge_id: edgeId }
}

/**
 * Multi-hop traversal from a starting fact.
 * Answers 'what is connected to X?' across the entity graph.
 */
export function registerGraphTraverseTool(
  api: OpenClawPluginApi,
  qdrantClient: QdrantClient,
  _cfg: NexusConfig,
  toolName = "nexus_graph_traverse",
): void {
  api.registerTool(
    {
      name: toolName,
      label: "Nexus Graph Traverse",
      description:
        "Knowledge Graph: Multi-hop traversal from a starting fact. " +
        "Answers 'what is connected to X?' across the entity graph. " +
        "Returns a list of { fact_id, depth, relation, path }.",
      parameters: Type.Object({
        fact_id: Type.String({ description: "The Qdrant point ID to start traversal from" }),
        max_depth: Type.Optional(
          Type.Number({
            description: "Maximum hops (default 3)",
            default: 3,
            minimum: 1,
            maximum: TRAVERSE_DEPTH_MAX,
          }),
        ),
        relation: Type.Optional(
          Type.String({
            description: "Only follow edges with this relation (e.g. 'manages', 'runs_on')",
          }),
        ),
        target_type: Type.Optional(
          Type.String({
            description: "Only return targets with this entity_type (e.g. 'device', 'service')",
          }),
        ),
      }),
      async execute(
        _toolCallId: string,
        params: { fact_id: string; max_depth?: number; relation?: string; target_type?: string },
      ) {
        const { fact_id: factId } = params
        // H123: bound the BFS fan-out (raw host value never reaches the loop).
        const maxDepth = clampInt(params.max_depth, TRAVERSE_DEPTH_DEFAULT, 1, TRAVERSE_DEPTH_MAX)
        const relation = params.relation || undefined
        const targetType = params.target_type || undefined

        try {
          const point = await qdrantClient.scrollPoint(factId)
          if (!point) {
            return {
              content: [{ type: "text" as const, text: `Fact ${factId} not found` }],
            }
          }

          // BFS traversal over edges in Qdrant payloads (shared lib/bfs.ts).
          const { steps: results } = await bfsEdges(qdrantClient, factId, maxDepth, {
            relation,
            targetType,
          })

          if (results.length === 0) {
            return {
              content: [{ type: "text" as const, text: "No connected facts found." }],
            }
          }

          const text = results
            .map((r, i) => {
              const pathStr = (r.path as string[]).join(" → ")
              return `${i + 1}. [depth=${r.depth}] ${r.relation} → ${r.fact_id} (path: ${pathStr})`
            })
            .join("\n")

          return {
            content: [
              { type: "text" as const, text: `Found ${results.length} connected facts:\n\n${text}` },
            ],
            details: { count: results.length, results },
          }
        } catch (err) {
          log.error("graph_traverse failed", err)
          return {
            content: [
              {
                type: "text" as const,
                text: `Graph traverse failed: ${err instanceof Error ? err.message : String(err)}`,
              },
            ],
          }
        }
      },
    },
    { name: toolName },
  )
}

/**
 * Find all entity-typed memories in Qdrant.
 * Returns a list of { id, name, entity_type, content, attributes }.
 */
export function registerFindEntitiesTool(
  api: OpenClawPluginApi,
  qdrantClient: QdrantClient,
  _cfg: NexusConfig,
  toolName = "nexus_find_entities",
): void {
  api.registerTool(
    {
      name: toolName,
      label: "Nexus Find Entities",
      description:
        "Knowledge Graph: Find all entity-typed memories in Qdrant. " +
        "Returns a list of { id, name, entity_type, content, attributes }.",
      parameters: Type.Object({
        entity_type: Type.Optional(
          Type.String({
            description:
              "Filter by entity type: device, service, person, location, organization, concept, software, protocol",
          }),
        ),
        limit: Type.Optional(
          Type.Number({
            description: "Max results (default 50)",
            default: 50,
            minimum: 1,
            maximum: ENTITIES_LIMIT_MAX,
          }),
        ),
      }),
      async execute(
        _toolCallId: string,
        params: { entity_type?: string; limit?: number },
      ) {
        const entityType = params.entity_type || undefined
        // H123: bound the paginated scroll fan-out (per page).
        const limit = clampInt(params.limit, ENTITIES_LIMIT_DEFAULT, 1, ENTITIES_LIMIT_MAX)

        try {
          const filter: Record<string, unknown> = {
            must: [{ key: "category", match: { value: "entity" } }],
          }
          if (entityType) {
            ;(filter.must as Array<Record<string, unknown>>).push({
              key: "entity_type",
              match: { value: entityType },
            })
          }

          // W31-18(b): `limit` is the PER-PAGE scroll limit; scrollFiltered
          // walks up to 5 pages, so without a total cap this returned up to
          // limit*5 entities. Pass `limit` as the total cap as well.
          const points = await qdrantClient.scrollFiltered(filter, limit, limit)
          const entities = points.map((pt) => {
            const payload = (pt.payload ?? {}) as Record<string, unknown>
            return {
              id: pt.id,
              name: payload.entity_name ?? "",
              entity_type: payload.entity_type ?? "",
              content: String(payload.content ?? "").slice(0, 200),
              attributes: payload.entity_attributes ?? {},
            }
          })

          if (entities.length === 0) {
            return {
              content: [{ type: "text" as const, text: "No entities found." }],
            }
          }

          const text = entities
            .map(
              (e, i) =>
                `${i + 1}. [${e.entity_type}] ${e.name} (id: ${e.id}) — ${e.content.slice(0, 80)}`,
            )
            .join("\n")

          return {
            content: [
              { type: "text" as const, text: `Found ${entities.length} entities:\n\n${text}` },
            ],
            details: { count: entities.length, entities },
          }
        } catch (err) {
          log.error("find_entities failed", err)
          return {
            content: [
              {
                type: "text" as const,
                text: `Find entities failed: ${err instanceof Error ? err.message : String(err)}`,
              },
            ],
          }
        }
      },
    },
    { name: toolName },
  )
}

/**
 * Get a subgraph centered on a fact for visualization.
 * Returns { nodes, edges } where nodes have { id, depth } and edges have { source, target, relation }.
 */
export function registerGetSubgraphTool(
  api: OpenClawPluginApi,
  qdrantClient: QdrantClient,
  cfg: NexusConfig,
  toolName = "nexus_get_subgraph",
): void {
  api.registerTool(
    {
      name: toolName,
      label: "Nexus Get Subgraph",
      description:
        "Knowledge Graph: Get a subgraph centered on a fact for visualization. " +
        "Returns { nodes, edges } where nodes have { id, depth } and edges have { source, target, relation }.",
      parameters: Type.Object({
        fact_id: Type.String({
          description: "The Qdrant point ID to center the subgraph on",
        }),
        max_depth: Type.Optional(
          Type.Number({
            description: "Maximum hops (default 2)",
            default: 2,
            minimum: 1,
            maximum: SUBGRAPH_DEPTH_MAX,
          }),
        ),
      }),
      async execute(_toolCallId: string, params: { fact_id: string; max_depth?: number }) {
        const { fact_id: factId } = params
        // H123: bound the BFS fan-out (raw host value never reaches the loop).
        const maxDepth = clampInt(params.max_depth, SUBGRAPH_DEPTH_DEFAULT, 1, SUBGRAPH_DEPTH_MAX)

        try {
          // Shared BFS (lib/bfs.ts), subgraph mode: collect nodes + edges.
          const { nodes, edges } = await bfsEdges(qdrantClient, factId, maxDepth, {
            collectEdges: true,
          })

          if (nodes.length <= 1) {
            return {
              content: [{ type: "text" as const, text: "No subgraph found around this fact." }],
            }
          }

          return {
            content: [
              {
                type: "text" as const,
                text: `Subgraph: ${nodes.length} nodes, ${edges.length} edges (centered on ${factId})`,
              },
            ],
            details: { nodes, edges },
          }
        } catch (err) {
          log.error("get_subgraph failed", err)
          return {
            content: [
              {
                type: "text" as const,
                text: `Get subgraph failed: ${err instanceof Error ? err.message : String(err)}`,
              },
            ],
          }
        }
      },
    },
    { name: toolName },
  )
}

/**
 * Get directly related facts (1-hop, bidirectional).
 * Returns a list of { fact_id, relation, edge_id, direction }.
 */
export function registerGetRelatedTool(
  api: OpenClawPluginApi,
  qdrantClient: QdrantClient,
  _cfg: NexusConfig,
  toolName = "nexus_get_related",
): void {
  api.registerTool(
    {
      name: toolName,
      label: "Nexus Get Related",
      description:
        "Knowledge Graph: Get directly related facts (1-hop, bidirectional). " +
        "Returns a list of { fact_id, relation, edge_id, direction }.",
      parameters: Type.Object({
        fact_id: Type.String({
          description: "The Qdrant point ID to find neighbors for",
        }),
        relation: Type.Optional(
          Type.String({
            description: "Only return edges with this relation (e.g. 'manages')",
          }),
        ),
      }),
      async execute(_toolCallId: string, params: { fact_id: string; relation?: string }) {
        const { fact_id: factId } = params
        const relation = params.relation || undefined

        try {
          const point = await qdrantClient.scrollPoint(factId)
          if (!point) {
            return {
              content: [{ type: "text" as const, text: `Fact ${factId} not found` }],
            }
          }

          const edges = (point.payload?.edges ?? []) as Array<Record<string, unknown>>
          const results: Array<Record<string, unknown>> = []

          // Outgoing edges
          for (const edge of edges) {
            if (!isActiveEdge(edge)) continue

            // H125: only validated values may enter the public contract —
            // results used to carry `fact_id: undefined` / `edge_id: undefined`.
            const normalized = normalizeEdge(edge)
            if (!normalized) {
              log.debug(`get_related: skipping malformed edge on ${factId}`)
              continue
            }
            if (relation && normalized.relation !== relation) continue

            results.push({
              fact_id: normalized.targetId,
              relation: normalized.relation,
              edge_id: normalized.edgeId,
              direction: "outgoing",
            })
          }

          // Incoming edges
          const incoming = await qdrantClient.findIncomingEdges(factId, relation)
          for (const edge of incoming) {
            // W31-18(a): run the SAME validation gate as the outgoing branch
            // (normalizeEdge/H125). Raw casts used to leak `fact_id:
            // undefined` / `edge_id: undefined` into the public contract.
            const normalized = validateIncomingEdge(edge)
            if (!normalized) {
              log.debug(`get_related: skipping malformed incoming edge on ${factId}`)
              continue
            }
            if (relation && normalized.relation !== relation) continue
            results.push({
              fact_id: normalized.fact_id,
              relation: normalized.relation,
              edge_id: normalized.edge_id,
              direction: "incoming",
            })
          }

          if (results.length === 0) {
            return {
              content: [{ type: "text" as const, text: "No related facts found." }],
            }
          }

          const text = results
            .map((r, i) => {
              const dir = r.direction === "outgoing" ? "→" : "←"
              return `${i + 1}. ${dir} [${r.relation}] ${r.fact_id} (${r.direction})`
            })
            .join("\n")

          return {
            content: [
              { type: "text" as const, text: `Found ${results.length} related facts:\n\n${text}` },
            ],
            details: { count: results.length, results },
          }
        } catch (err) {
          log.error("get_related failed", err)
          return {
            content: [
              {
                type: "text" as const,
                text: `Get related failed: ${err instanceof Error ? err.message : String(err)}`,
              },
            ],
          }
        }
      },
    },
    { name: toolName },
  )
}

File: plugins/claude-code/scripts/graph_traverse.py
#!/usr/bin/env python3
"""Claude Code Plugin — Knowledge Graph Traversal Script.

Provides graph traversal, entity search, subgraph, and related-facts queries
for Claude Code via PreToolExecution / hooks.

Usage:
    python3 graph_traverse.py --action traverse --fact-id <id> [--max-depth 3]
    python3 graph_traverse.py --action find_entities [--entity-type device] [--limit 50]
    python3 graph_traverse.py --action subgraph --fact-id <id> [--max-depth 2]
    python3 graph_traverse.py --action related --fact-id <id>

--max-depth defaults to 3 for traverse and 2 for subgraph; --entity-type also
acts as the target_type filter for traverse.
"""

import argparse
import json
import os
import sys

# Append (never prepend) nexus-memory to the module search path: a directory
# writable by the invoking user at sys.path[0] could shadow stdlib or
# third-party modules (import hijack).
sys.path.append(os.path.expanduser("~/nexus-memory"))

QDRANT_URL = os.environ.get("NEXUS_QDRANT_URL", "http://localhost:6333")
COLLECTION = os.environ.get("NEXUS_COLLECTION", "nexus")

# Upper bounds for user-supplied integers: an unbounded depth walks the whole
# graph, a negative limit is forwarded to the Qdrant scroll.
MAX_DEPTH_LIMIT = 10
LIMIT_MAX = 1000


def _emit_error(message: str) -> None:
    """Print the single error schema to stdout (pinned stdout contract, H1).

    Every failure path uses this shape so callers can branch on ``status``
    regardless of whether validation or the backend failed.
    """
    print(json.dumps({"status": "error", "error": message}))


def traverse(fact_id: str, max_depth: int = 3, relation: str = "", target_type: str = "") -> dict:
    """Multi-hop BFS traversal from a starting fact."""
    from nexus.graph.graph import SkillGraph
    from nexus.graph.traversal import GraphTraversal

    sg = SkillGraph(qdrant_url=QDRANT_URL, collection=COLLECTION)
    try:
        sg.initialize()
        gt = GraphTraversal(sg)
        results = gt.traverse(
            fact_id,
            max_depth=max_depth,
            relation=relation or None,
            target_type=target_type or None,
        )
        return {"results": results}
    finally:
        sg.store.close()


def find_entities(entity_type: str = "", limit: int = 50) -> dict:
    """Find all entity-typed memories."""
    from nexus.graph.graph import SkillGraph
    from nexus.graph.traversal import GraphTraversal

    sg = SkillGraph(qdrant_url=QDRANT_URL, collection=COLLECTION)
    try:
        sg.initialize()
        gt = GraphTraversal(sg)
        results = gt.find_entities(
            entity_type=entity_type or None,
            limit=limit,
        )
        return {"entities": results}
    finally:
        sg.store.close()


def get_subgraph(fact_id: str, max_depth: int = 2) -> dict:
    """Get a subgraph centered on a fact."""
    from nexus.graph.graph import SkillGraph
    from nexus.graph.traversal import GraphTraversal

    sg = SkillGraph(qdrant_url=QDRANT_URL, collection=COLLECTION)
    try:
        sg.initialize()
        gt = GraphTraversal(sg)
        result = gt.get_subgraph(fact_id, max_depth=max_depth)
        return result
    finally:
        sg.store.close()


def get_related(fact_id: str, relation: str = "") -> dict:
    """Get directly related facts (1-hop)."""
    from nexus.graph.graph import SkillGraph
    from nexus.graph.traversal import GraphTraversal

    sg = SkillGraph(qdrant_url=QDRANT_URL, collection=COLLECTION)
    try:
        sg.initialize()
        gt = GraphTraversal(sg)
        results = gt.get_related(fact_id, relation=relation or None)
        return {"results": results}
    finally:
        sg.store.close()


def main():
    parser = argparse.ArgumentParser(description="Nexus Memory Knowledge Graph")
    parser.add_argument("--action", required=True,
                        choices=["traverse", "find_entities", "subgraph", "related"],
                        help="Query action to perform")
    parser.add_argument("--fact-id", default="", help="Qdrant point ID (for traverse/subgraph/related)")
    parser.add_argument("--max-depth", type=int, default=None,
                        help=f"Maximum hops 1-{MAX_DEPTH_LIMIT} (default: 3 for traverse, 2 for subgraph)")
    parser.add_argument("--relation", default="", help="Filter by relation type")
    parser.add_argument("--entity-type", default="",
                        help="Filter by entity type (find_entities; for traverse it filters target_type)")
    parser.add_argument("--limit", type=int, default=50,
                        help=f"Max results for find_entities (1-{LIMIT_MAX})")
    args = parser.parse_args()

    # Validate ranges before they reach the traversal / Qdrant scroll layer.
    if args.max_depth is not None and not 1 <= args.max_depth <= MAX_DEPTH_LIMIT:
        _emit_error(f"--max-depth must be between 1 and {MAX_DEPTH_LIMIT}")
        sys.exit(1)
    if not 1 <= args.limit <= LIMIT_MAX:
        _emit_error(f"--limit must be between 1 and {LIMIT_MAX}")
        sys.exit(1)

    try:
        if args.action == "traverse":
            if not args.fact_id:
                _emit_error("fact-id required for traverse")
                sys.exit(1)
            result = traverse(args.fact_id, args.max_depth if args.max_depth is not None else 3,
                              args.relation, args.entity_type)
        elif args.action == "find_entities":
            result = find_entities(args.entity_type, args.limit)
        elif args.action == "subgraph":
            if not args.fact_id:
                _emit_error("fact-id required for subgraph")
                sys.exit(1)
            # Only forward an explicit --max-depth: get_subgraph()'s own
            # default (2) must win when the flag was omitted (the shared
            # parser default of 3 walked one hop deeper than documented).
            if args.max_depth is None:
                result = get_subgraph(args.fact_id)
            else:
                result = get_subgraph(args.fact_id, args.max_depth)
        elif args.action == "related":
            if not args.fact_id:
                _emit_error("fact-id required for related")
                sys.exit(1)
            result = get_related(args.fact_id, args.relation)
        else:
            _emit_error(f"Unknown action: {args.action}")
            sys.exit(1)

        print(json.dumps(result, indent=2))
    except Exception as e:
        # Include the exception type: an opaque str(e) made backend failures
        # (Qdrant/connection) indistinguishable from programming errors.
        _emit_error(f"{type(e).__name__}: {e}")
        sys.exit(1)


if __name__ == "__main__":
    main()